File: union_find.c


int* InitUnionFind(int n) {
    int* set = (int*)malloc(n*sizeof(int));
    for (int i = 0; i < n; i++) {
        set[i] = i;
    }
    return set;
}

int Find(int* set, int p) {
    if (set[p] == p) {
        return p;
    }
    int root = Find(set, set[p]);
    set[p] = root;
    return root;
}

int Union(int* set, int p, int q) {
    int root_p = Find(set, p);
    int root_q = Find(set, q);
    set[root_p] = root_q;
    return root_q;
}

File: union_find.java

class UnionFind {
    private int[] s;
    private int cnt;
    public UnionFind(int size) {
        s = new int[size];
        for (int i = 0; i < size; i++) {
            s[i] = -1;
        }
        cnt = size;
    }

    public int union(int a, int b) {
        int ra = find(a);
        int rb = find(b);
        if (ra != rb) {
            s[ra] = rb;
            cnt -= 1;
        }
        return rb;
    }

    public int find(int a) {
        if (s[a] == -1) {
            return a;
        }
        int root = find(s[a]);
        s[a] = root;
        return root;
    }

    public int count() {
        return cnt;
    }
}